corim/unsignedcorim.go

```diff
@@ -4,104 +4,219 @@
 package corim
 
 import (
+	"encoding/json"
+	"errors"
 	"fmt"
 
 	"github.com/veraison/corim/comid"
 	"github.com/veraison/eat"
 	"github.com/veraison/swid"
 )
 
+// UnsignedCorim is the top-level representation of the unsigned-corim-map with
+// CBOR and JSON serialization.
 type UnsignedCorim struct {
-	ID            ID             `cbor:"0,keyasint" json:"corim-id"`
+	ID            swid.TagID     `cbor:"0,keyasint" json:"corim-id"`
 	Tags          []Tag          `cbor:"1,keyasint" json:"tags"`
 	DependentRims *[]Locator     `cbor:"2,keyasint,omitempty" json:"dependent-rims,omitempty"`
 	Profiles      *[]eat.Profile `cbor:"3,keyasint,omitempty" json:"profiles,omitempty"`
 }
 
-type ID struct {
-	val interface{}
+// NewUnsignedCorim instantiates an empty UnsignedCorim
+func NewUnsignedCorim() *UnsignedCorim {
+	return &UnsignedCorim{}
 }
 
-func (o ID) MarshalCBOR() ([]byte, error) {
-	return em.Marshal(o.val)
+// SetID sets the corim-id in the unsigned-corim-map to the supplied value.  The
+// corim-id can be passed as UUID in string or binary form (i.e., byte array),
+// or as a (non-empty) string
+func (o *UnsignedCorim) SetID(v interface{}) *UnsignedCorim {
+	if o != nil {
+		tagID := swid.NewTagID(v)
+		if tagID == nil {
+			return nil
+		}
+		o.ID = *tagID
+	}
+	return o
 }
 
-func (o *ID) UnmarshalCBOR(data []byte) error {
-	var s string
+// GetID retrieves the corim-id from the unsigned-corim-map as a string
+func (o UnsignedCorim) GetID() string {
+	return o.ID.String()
+}
 
-	if dm.Unmarshal(data, &s) == nil {
-		o.val = s
-		return nil
-	}
+// AddComid appends the CBOR encoded (and appropriately tagged) CoMID to the
+// tags array of the unsigned-corim-map
+func (o *UnsignedCorim) AddComid(c comid.Comid) *UnsignedCorim {
+	if o != nil {
+		if c.Valid() != nil {
+			return nil
+		}
+
+		comidCBOR, err := c.ToCBOR()
+		if err != nil {
+			return nil
+		}
 
-	var u comid.UUID
+		// d9 01fa # tag(506)
+		comidTag := []byte{0xd9, 0x01, 0xfa}
 
-	if dm.Unmarshal(data, &u) == nil {
-		o.val = u
-		return nil
-	}
+		taggedComid := append(comidTag, comidCBOR...)
 
-	return fmt.Errorf("unknown corim-id type (CBOR: %x)", data)
+		o.Tags = append(o.Tags, taggedComid)
+	}
+	return o
 }
 
-type Tag []byte
+// AddCoswid appends the CBOR encoded (and appropriately tagged) CoSWID to the
```

Comment:
Correct me if I am wrong but we should also have an Interface API to add a CBOR Encoded CoSWID directly into the CoRIM, rather than been passed as a structure as argument.

Reply:
Sure, maybe we could add that in the future.  For now, I have preferred a tighter integration with the rest of the Veraison ecosystem.

Reply:
Future work is fine with me, so long as we track it via github issue. Also in Veraison PSA, perhaps we will have all the FW Measurements as Ref Values as part of one or more CoMID ?

Reply:
Also thinking a bit more, if a CoRIM Creator is a different entity than a CoMID & CoSWID generators, then we should extend the same flexibility for CoMID also.